### PR TITLE
Sv2 int channel deletion

### PR DESCRIPTION
**Pull request by dobertRowneySr**
Refers to issues #2630 #2640 
Replacing previous PR, since it is more convenient for me to create a new one

### CODE REVIEW

File: runtime-modules/content/src/lib.rs [hunk 1]
```diff
@@ -237,38 +235,44 @@ pub struct ChannelOwnershipTransferRequestRecord<
...
+type ChannelCreationParameters<T> =
+    ChannelCreationParameters_<NewAsset<T>, <T as frame_system::Trait>::AccountId>;
+
 /// Information about channel being updated.
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
-pub struct ChannelUpdateParameters<ContentParameters, AccountId> {
-    /// Assets referenced by metadata
-    assets: Option<Vec<NewAsset<ContentParameters>>>,
+pub struct ChannelUpdateParameters_<NewAsset, AccountId> {
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs]: We use `Record` postfix for this purpose: `ChannelUpdateParametersRecord`

File: runtime-modules/content/src/lib.rs [hunk 2]
```diff
@@ -296,34 +300,64 @@ pub struct VideoCategoryUpdateParameters {
...
     new_meta: Option<Vec<u8>>,
 }
 
+type VideoUpdateParameters<T> = VideoUpdateParameters_<NewAsset<T>>;
+
 /// A video which belongs to a channel. A video may be part of a series or playlist.
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
-pub struct Video<ChannelId, SeriesId> {
+pub struct Video_<ChannelId, SeriesId, DataObjectId: Ord> {
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs]: Please, use `Record` postfix here instead of `_`

File: runtime-modules/content/src/lib.rs [hunk 3]
```diff
@@ -699,20 +750,30 @@ decl_module! {
...
-                // check assets can be uploaded to storage.
-                // update can_add_content() to only take &refrences
-                T::StorageSystem::can_add_content(
-                    object_owner.clone(),
-                    upload_parameters.clone(),
-                )?;
-
-                Some((upload_parameters, object_owner))
+            // Maybe add specified assets and get the actual number assets added
+            let maybe_num_assets = if let Some(assets) = &params.assets {
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs]: This code snippet could be simplified. Please, get rid of the nested `if-else` using `map` and `unwrap_*` methods.

File: runtime-modules/content/src/lib.rs [hunk 4]
```diff
@@ -760,13 +827,35 @@ decl_module! {
...
+                Self::bag_id_for_channel(&channel_id),
+                assets.clone(),
+            )?;
 
             //
             // == MUTATION SAFE ==
             //
 
-            T::StorageSystem::atomically_remove_content(&object_owner, &assets)?;
+            // update onchain channel status
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs:838]: Please, use `mutate` syntax instead.

File: runtime-modules/content/src/permissions/mod.rs [hunk 5]
```diff
@@ -5,7 +5,7 @@ pub use curator_group::*;
 pub use crate::errors::*;
 use crate::*;
 pub use codec::{Codec, Decode, Encode};
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/permissions/mod.rs:7]: Remove commented code.

File: runtime-modules/content/src/permissions/mod.rs [hunk 6]
```diff
@@ -15,7 +15,7 @@ use sp_runtime::traits::{MaybeSerializeDeserialize, Member};
 // use frame_system::ensure_root;
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/permissions/mod.rs:14]: Remove commented code.

File: runtime-modules/common/src/storage.rs [hunk 7]
```diff
@@ -33,16 +24,16 @@ impl<MemberId, ChannelId, DAOId> Default for StorageObjectOwner<MemberId, Channe
 // To be implemented by current storage data_directory runtime module.
 // Defined in 'common' package
 pub trait StorageSystem<T: crate::StorageOwnership + crate::MembershipTypes> {
-    fn atomically_add_content(
-        owner: StorageObjectOwner<T::MemberId, T::ChannelId, T::DAOId>,
-        content_parameters: Vec<ContentParameters<T::ContentId, T::DataObjectTypeId>>,
-    ) -> DispatchResult;
+    // fn atomically_add_content(
```
Comment (shamil-gadelshin) [inline runtime-modules/common/src/storage.rs]: Please, remove commented code.

File: runtime-modules/content/src/lib.rs [hunk 8]
```diff
@@ -78,16 +74,22 @@ impl NumericIdentifier for u64 {}
...
-    + StorageOwnership
-    + MembershipTypes
-    + GovernanceCurrency
+    frame_system::Trait + ContentActorAuthenticator + Clone + GovernanceCurrency + storage::Trait
 {
     /// The overarching event type.
     type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
 
+    /// DAO id representation.
+    type DAOId: Parameter
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs]: Shouldn't we rename it to `DaoId` to respect `clippy` complains?

File: runtime-modules/content/src/lib.rs [hunk 9]
```diff
@@ -638,45 +672,62 @@ decl_module! {
...
-            params: ChannelCreationParameters<ContentParameters<T>, T::AccountId>,
+            params: ChannelCreationParameters<T>,
         ) {
             ensure_actor_authorized_to_create_channel::<T>(
-                origin,
+                origin.clone(),
                 &actor,
             )?;
 
+        // channel creator account
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs]: please, format the code.

File: runtime-modules/content/src/lib.rs [hunk 10]
```diff
@@ -638,45 +672,62 @@ decl_module! {
...
 
-            let object_owner = StorageObjectOwner::<T>::Channel(channel_id);
+            // get uploading parameters if assets have to be saved on storage
+            let maybe_upload_parameters = Self::pick_upload_parameters_from_assets(
+                &params.assets,
+                &channel_id
+            );
+
+            // number of assets to be uploaded
+            let num_assets: u64 = if let Some(upload_parameters) = maybe_upload_parameters{
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs]: please, simplify the code:
- `saturated_into` will get rid of the error
- `unwrap_*` could help with `else 0`

File: runtime-modules/content/src/lib.rs (hunk 3 above)
Comment (shamil-gadelshin) [reply runtime-modules/content/src/lib.rs]: Also, use `saturated_into` to get rid of the conversion errors

File: runtime-modules/content/src/lib.rs [hunk 11]
```diff
@@ -912,41 +1001,56 @@ decl_module! {
...
+
+            // if storaged uploading is required save the object id for the video
+            let maybe_data_object_ids = if let Some(upload_parameters) =
+                maybe_upload_parameters {
+                    let beginning_id = Storage::<T>::next_data_object_id();
+                    Storage::<T>::upload_data_objects(upload_parameters)?;
+                    let ending_id = Storage::<T>::next_data_object_id();
+                    let mut data_objects_id_set:
+                        BTreeSet<<T as storage::Trait>::DataObjectId> = BTreeSet::new();
+                    for el in beginning_id..=ending_id {
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs]: Why do we use inclusive range here?

File: runtime-modules/content/src/lib.rs [hunk 12]
```diff
@@ -912,41 +1001,56 @@ decl_module! {
...
-                object_owner,
-                content_parameters,
-            )?;
+            // adding the content to storage node if uploading is needed
+            let maybe_upload_parameters = Self::pick_upload_parameters_from_assets(
+                &params.assets,
+                &channel_id
+            );
+
+            // if storaged uploading is required save the object id for the video
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs]: Please, simplify the code.

File: runtime-modules/content/src/lib.rs [hunk 13]
```diff
@@ -1031,13 +1139,10 @@ decl_module! {
...
             VideoById::<T>::remove(video_id);
 
-            // Update corresponding channel
-            // Remove recently deleted video from the channel
-            ChannelById::<T>::mutate(channel_id, |channel| {
-                if let Some(index) = channel.videos.iter().position(|x| *x == video_id) {
-                    channel.videos.remove(index);
-                }
-            });
+            // Decrease video count for the channel
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs:1127]: use `mutate` syntax instead

File: runtime-modules/content/src/lib.rs [hunk 14]
```diff
@@ -1360,6 +1474,12 @@ impl<T: Trait> Module<T> {
         }
     }
 
+    fn bag_id_for_channel(channel_id: &T::ChannelId) -> storage::BagId<T> {
+        // retrieve bag id from channel id
+        let dyn_bag = DynamicBagIdType::<T::MemberId, T::ChannelId>::Channel(channel_id.clone());
+        BagIdType::<T::MemberId, T::ChannelId>::Dynamic(dyn_bag)
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/lib.rs]: `From` trait is implemented for this conversion.

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
Issues found:
- multiple build errors
- multiple clippy warnings
- no tests for added code
- merge conflicts
- please format code inside extrinsics as well
- remove all commented code
- change comments to doc-comments (especially for structs)
- use `Record` postfix instead of `_`
- use `saturated_into` conversion to avoid errors
- avoid using patterns `else None` and `else 0` for optional types - they could be rewritten using `map` or `unwrap_*` methods
- add or change comments for affected extrinsics in doc-comments format.
- remove non-implemented extrinsics or implement them

File: runtime-modules/content/src/lib.rs [hunk 15]
```diff
@@ -237,38 +235,44 @@ pub struct ChannelOwnershipTransferRequestRecord<
...
 >;
 
 /// Information about channel being created.
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
-pub struct ChannelCreationParameters<ContentParameters, AccountId> {
+pub struct ChannelCreationParameters_<NewAsset, AccountId> {
     /// Assets referenced by metadata
-    assets: Vec<NewAsset<ContentParameters>>,
+    assets: NewAsset,
```
Comment (Lezek123) [inline runtime-modules/content/src/lib.rs]: I found this change to be quite confusing. There are multiple assets associated with the channel (currently `coverPhoto` and `avatarPhoto`), but the type is called `NewAsset` (in a singular form).

Also, `NewAsset` enum consists of 2 variants:
- `Urls(Vec<Url>) // Multiple url strings pointing at an asset` - which suggests urls only for one asset can be provided
- `Upload(CreationUploadParameters<AccountId, Balance>)` - which can be used to upload multiple assets

I would really appreciate if the `NewAsset` type was more consistent (I think the previous `assets: Vec<NewAsset>` approach was much clearer)

File: runtime-modules/content/src/lib.rs [hunk 16]
```diff
@@ -296,34 +300,64 @@ pub struct VideoCategoryUpdateParameters {
...
+    /// Expected data size fee value for this extrinsic call.
+    pub expected_data_size_fee: Balance,
+}
+
+/// Information about the video being created.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub struct VideoCreationParameters_<NewAsset> {
+    /// Content creation parameters
+    assets: NewAsset,
```
Comment (Lezek123) [inline runtime-modules/content/src/lib.rs]: Same issue as with `ChannelCreationParameters.assets`

File: runtime-modules/content/src/lib.rs [hunk 17]
```diff
@@ -296,34 +300,64 @@ pub struct VideoCategoryUpdateParameters {
 /// Information about the video being created.
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
-pub struct VideoCreationParameters<ContentParameters> {
-    /// Assets referenced by metadata
-    assets: Vec<NewAsset<ContentParameters>>,
+/// Information regarding the content being created
+pub struct CreationUploadParameters<AccountId, Balance> {
+    /// Public key used authentication in upload to liason.
+    pub authentication_key: Vec<u8>,
```
Comment (Lezek123) [inline runtime-modules/content/src/lib.rs]: I think this was dropped in https://github.com/Joystream/joystream/pull/2611

**Review by Lezek123 — COMMENTED** [review]
I'm not an expert on the runtime, but I spotted a few changes that got me a little confused.

File: runtime-modules/content/src/lib.rs (hunk 11 above)
Comment (dobertRowneySr) [reply runtime-modules/content/src/lib.rs]: my mistake. fixed

File: runtime-modules/content/src/lib.rs (hunk 10 above)
Comment (dobertRowneySr) [reply runtime-modules/content/src/lib.rs]: done

File: runtime-modules/content/src/lib.rs (hunk 13 above)
Comment (dobertRowneySr) [reply runtime-modules/content/src/lib.rs:1127]: done

File: runtime-modules/content/src/lib.rs (hunk 14 above)
Comment (dobertRowneySr) [reply runtime-modules/content/src/lib.rs]: done

File: runtime-modules/content/src/lib.rs (hunk 8 above)
Comment (dobertRowneySr) [reply runtime-modules/content/src/lib.rs]: clippy doesn't complain on my machine

Comment (dobertRowneySr) [reply runtime-modules/content/src/lib.rs]: removed DAO references btw. As requested in the Giza release plan

File: runtime/src/lib.rs [hunk 18]
```diff
@@ -443,34 +443,33 @@ impl content::Trait for Runtime {
     type SeriesId = SeriesId;
     type ChannelOwnershipTransferRequestId = ChannelOwnershipTransferRequestId;
     type MaxNumberOfCuratorsPerGroup = MaxNumberOfCuratorsPerGroup;
-    type StorageSystem = (); // TODO: Add storage integration
 }
 
 // TODO: Remove after the integration with the Content pallet.
```
Comment (shamil-gadelshin) [inline runtime/src/lib.rs]: This pull request IS this integration, isn't it?

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
Looks good. There are several code comments issues.

File: runtime-modules/content/src/permissions/mod.rs (hunk 6 above)
Comment (shamil-gadelshin) [reply runtime-modules/content/src/permissions/mod.rs:14]: Here.

File: runtime-modules/content/src/permissions/mod.rs (hunk 5 above)
Comment (shamil-gadelshin) [reply runtime-modules/content/src/permissions/mod.rs:7]: Here.

File: runtime/src/lib.rs (hunk 18 above)
Comment (shamil-gadelshin) [reply runtime/src/lib.rs]: Should we remove `impl common::storage::StorageSystem<Runtime> for ()` ?

Comment (ignazio-bovo) [reply runtime/src/lib.rs:448]: I didn't understand the first question but this PR entails:
- #2574 
- #2465 
- storage integration

Comment (dobertRowneySr) [reply runtime/src/lib.rs]: I agree. My guess is that StorageSystem was the common trait used before the v2 version

**Review by shamil-gadelshin — APPROVED** [review]
LGTM

**Pull request merged**